gossip: move to gRPC

tamird opened the pull request:
This depends on https://github.com/grpc/grpc-go/pull/514, but it's ready for review.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4080)

<!-- Reviewable:end -->

tamird committed:
build: add gRPC codegen and hook up logging

tamird committed:
gossip: move to gRPC

tamird committed:
gossip: set node ID before starting gossip

This test is nonsense. It relies on the long-polling behaviour of the
client (which will start a getGossip immediately after the previous
getGossip returns) and the not-long-polling behaviour of
fakeGossipServer, which doesn't block until new deltas are available,
and just returns instantly.

tamird committed:
leaktest: whitelist grpc server

tamird committed:
server: better errors

tschottdorf commented:
Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 15 of 15 files at r3.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 64 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XmuP78X5vgUevtGam:-K9XmuP81Ff5y6WmgMfM:1617912550) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L64)):*
tripped over the `it` here,  maybe s/it/itself/?

---

*[gossip/client.go, line 114 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XopwXR_OKIZCEO_FZ:-K9XopwXR_OKIZCEO_F_:2081457987) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L114)):*
looks odd at first. Reading where it's used it doesn't look as bad, but add a comment.

---

*[gossip/client.go, line 288 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xq3ZqzAW3OIeZaFhm:-K9Xq3ZqzAW3OIeZaFhn:-1597385690) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L288)):*
s/is sent/sends/

---

*[gossip/client.go, line 289 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xq8EGgGWeSnEIzRkx:-K9Xq8EGgGWeSnEIzRky:-1341251074) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L289)):*
s/server/remote/

---

*[gossip/client.go, line 302 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XpegzS8dG5s8noo_x:-K9XpegzS8dG5s8noo_y:-534544231) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L302)):*
maybe `handleGossip` ought to be called `handleResponse`.

---

*[gossip/server.go, line 62 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XrET98JaANe9txfXX:-K9XrETAcWZ_TzxONC8Y:-2054210632) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L62)):*
Add a brief comment here as well.

---

*[gossip/server.go, line 100 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XriEtZRsvz5aOUQUn:-K9XriEtZRsvz5aOUQUo:1947178808) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L100)):*
s/connet/connect/

---

*[gossip/server.go, line 134 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XsGoOcPpvAKGNfjFv:-K9XsGoOcPpvAKGNfjFw:2040715879) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L134)):*
wouldn't this check also catch the above duplicate connection check? Maybe this can just go above it and run in either case (assuming the hash lookup vs. full loop doesn't matter).

---

*[gossip/server.go, line 140 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xsf3r-0EokQJRyxSL:-K9Xsf3r-0EokQJRyxSM:-700756099) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L140)):*
s/is sends/sends/

---

*[gossip/server.go, line 146 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XsjKBdlgkkgSFXIg4:-K9XsjKBdlgkkgSFXIg5:-714926407) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L146)):*
why use `new` here?

---

*[gossip/server.go, line 156 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xu9wMn-qH9BDWwW8U:-K9Xu9wMn-qH9BDWwW8V:-498397581) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L156)):*
isn't this a race (depending on how `s.ready` gets triggered)?

---

*[gossip/server.go, line 176 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuQpGq58DWBxSiC43:-K9XuQpHekRXEewNPn7s:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L176)):*
ditto

---

*[gossip/server.go, line 204 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuZAHjqodwXaGK13F:-K9XuZAHjqodwXaGK13G:-420988434) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L204)):*
why not

```go
var err error
args, err = stream.Recv()
```

---

*[gossip/server.go, line 268 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuiMfn2e8zx4z4naz:-K9XuiMfn2e8zx4z4nb-:2024135660) ([raw file](https://github.com/cockroachdb/cockroach/blob/554c23f1cba40c56f760a7dfbe2d6cf52a8c2cdc/gossip/server.go#L268)):*
good to see this stuff go.

---

*[server/node_test.go, line 246 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XmJsxEqs83hZwgheW:-K9XmJsynbW1W2EmQxSc:-65239310) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ac3b6236978c0e20cabf21f5631fb16ff68c8/server/node_test.go#L246)):*
hasn't been flaky as far as I know, but can't hurt to go up from `50ms` here.

---

*[server/node_test.go, line 67 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xuq1Ai9uwXZ1IehBk:-K9Xuq1Ai9uwXZ1IehBl:-523759396) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/server/node_test.go#L67)):*
is this a crutch? If so, add a comment.

---

*[server/server.go, line 287 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XvIHtCah5dZzMMrYh:-K9XvIHtCah5dZzMMrYi:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/server/server.go#L287)):*
ditto

---

*[util/net.go, line 139 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XvxZeJbjVbh2nBPnG:-K9XvxZeJbjVbh2nBPnH:210401286) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/util/net.go#L139)):*
why wasn't this needed before?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 64 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XmuP78X5vgUevtGam:-K9Y3e6WSjaCql2qGhf_:2102208419) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L64)):*
Updated, take another look.

---

*[gossip/client.go, line 114 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XopwXR_OKIZCEO_FZ:-K9Y5_5JY3OzD3UKdfjt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L114)):*
Done.

---

*[gossip/client.go, line 288 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xq3ZqzAW3OIeZaFhm:-K9Y5j1TPEFhhxtkNd4a:439824157) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L288)):*
This comment isn't useful, I just removed it.

---

*[gossip/client.go, line 289 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xq8EGgGWeSnEIzRkx:-K9Y5wgdJV_k01CLZIuT:63689955) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L289)):*
Removed the comment.

---

*[gossip/client.go, line 302 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XpegzS8dG5s8noo_x:-K9Y7Qu99Mjb2Sf8H3vt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L302)):*
Done.

---

*[gossip/server.go, line 62 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XrET98JaANe9txfXX:-K9Y65foAV5mMTOIhp54:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L62)):*
Done.

---

*[gossip/server.go, line 100 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XriEtZRsvz5aOUQUn:-K9Y67iKcyyWTEjgAcMp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L100)):*
Done.

---

*[gossip/server.go, line 134 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XsGoOcPpvAKGNfjFv:-K9Y6HPFOJmZEQeX87p6:1831815694) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L134)):*
Personally, I'd remove all the NodeID-based duplicate checking and always just see if we have a connection from this node (`nodeMap` would be inverted), but @spencerkimball objected to this, so he should weigh in here.

---

*[gossip/server.go, line 140 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xsf3r-0EokQJRyxSL:-K9Y75ym1Iusqth87u5p:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L140)):*
Done.

---

*[gossip/server.go, line 146 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XsjKBdlgkkgSFXIg4:-K9Y6o8nefaiDs2c7i6H:-1592974396) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L146)):*
No particular reason; just want a zero value on the heap.

---

*[gossip/server.go, line 156 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xu9wMn-qH9BDWwW8U:-K9Y6vPeGhiT6fS2YXYZ:569910924) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L156)):*
A race with what?

---

*[gossip/server.go, line 176 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuQpGq58DWBxSiC43:-K9Y7EmCvaapBOClEgPp:-64062301) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L176)):*
ditto - a race with what?

---

*[gossip/server.go, line 204 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuZAHjqodwXaGK13F:-K9Y7HzMtDSa0UgUjhTf:1470875955) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L204)):*
Remnant of old code that was manually manipulating the mutex. Done.

---

*[server/node_test.go, line 246 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XmJsxEqs83hZwgheW:-K9Y7sSS0JGw1qgfGZS_:1499495330) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ac3b6236978c0e20cabf21f5631fb16ff68c8/server/node_test.go#L246)):*
Any reason to do that in this change?

---

*[server/node_test.go, line 67 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xuq1Ai9uwXZ1IehBk:-K9Y7W30HA_OMuBrXywV:-1482797362) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/server/node_test.go#L67)):*
What do you mean by "crutch"? This is a partial recreation of gRPC's internal checks https://github.com/grpc/grpc-go/pull/514/files#diff-95e9a25b738459a2d3030e1e6fa2a718R61

I'll add a comment that points to that code when the PR is merged.

---

*[server/server.go, line 287 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XvIHtCah5dZzMMrYh:-K9Y7zmAAku-rf4huBBa:188755235) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/server/server.go#L287)):*
I've extracted this so I only have to add a comment in one place.

---

*[util/net.go, line 139 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XvxZeJbjVbh2nBPnG:-K9Y898HisWIAj7wggAH:-1677021394) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/util/net.go#L139)):*
Coincidence. `http2.ConfigureServer` mutates its argument's `TLSConfig`, and that's an essential part of the http2 connection handling. We need these modifications for `tls.NewListener` to do the right thing.

It just happened not to be needed before.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 64 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XmuP78X5vgUevtGam:-K9YCZyJgfVchID1arDh:-574926344) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L64)):*
s/Upoon/Upon/

---

*[gossip/client.go, line 301 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YCrryAaXgNbNiGPhe:-K9YCrryAaXgNbNiGPhf:1431733985) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d984c0c7375b7fc82c7f3398b95b88a75cc2ad1/gossip/client.go#L301)):*
s/passed to the outgoing RPC/attached to the stream/?

---

*[gossip/server.go, line 134 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XsGoOcPpvAKGNfjFv:-K9YDKq2O9fo_RUigksZ:-181034550) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L134)):*
yeah, if we can reliably filter out duplicate connections without recourse to NodeID, that seems preferable to me as well.

---

*[gossip/server.go, line 156 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xu9wMn-qH9BDWwW8U:-K9YDbFqEEcWSe23AcXG:1329547591) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L156)):*
ah, sorry; this is `RunTask`, not `RunAsyncTask`. Nevermind.

---

*[gossip/server.go, line 176 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuQpGq58DWBxSiC43:-K9YDmr_5Oz_QJrRCiZk:803797171) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L176)):*
this "ditto" was about `new` (but has been answered above).

---

*[server/node_test.go, line 246 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XmJsxEqs83hZwgheW:-K9YEAnbTJjMuxhPYXEt:160414568) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ac3b6236978c0e20cabf21f5631fb16ff68c8/server/node_test.go#L246)):*
no.

---

*[server/node_test.go, line 67 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Xuq1Ai9uwXZ1IehBk:-K9YE6T7UX-8OcuWDWvJ:686510732) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/server/node_test.go#L67)):*
I was just curious since this looked ad hoc.

---

*[util/net.go, line 185 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YFD59J5BpN4-5Qsvh:-K9YFD5ABe48CNtgVnZz:507310989) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d984c0c7375b7fc82c7f3398b95b88a75cc2ad1/util/net.go#L185)):*
doesn't this need a comment for the linters?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
LGTM

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-:-K9YFYJWWZSHJtbvbVjh:676834233)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 64 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XmuP78X5vgUevtGam:-K9YFXpQVtjT99gLnj0U:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/client.go#L64)):*
Done.

---

*[gossip/client.go, line 301 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YCrryAaXgNbNiGPhe:-K9YFYit91IEuPjNe8g2:-1942080066) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d984c0c7375b7fc82c7f3398b95b88a75cc2ad1/gossip/client.go#L301)):*
No, it's literally an argument that is passed.

---

*[util/net.go, line 185 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YFD59J5BpN4-5Qsvh:-K9YFon_xLHN-LkkRaLw:1707716069) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d984c0c7375b7fc82c7f3398b95b88a75cc2ad1/util/net.go#L185)):*
Ah yes, good eye. Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 13 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 301 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YCrryAaXgNbNiGPhe:-K9YIlzmqwUvf0NyN7d5:-997990693) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d984c0c7375b7fc82c7f3398b95b88a75cc2ad1/gossip/client.go#L301)):*
I'm confused by the comment. The below code would error out on `stream.Recv()`, right? And it would error since `Recv()` checks the context which wraps the stopper which has been used for constructing `stream`. I'm not sure where the "outgoing" RPC the comment mentions exists.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[util/net.go, line 190 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YJZdtQZgh3_4rCmN1:-K9YJZdtQZgh3_4rCmN2:262468467) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b289457e97178ef23f128a1cff0b6e80640fd0a/util/net.go#L190)):*
:thumbsup: 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 14 of 15 files reviewed at latest revision, 3 unresolved discussions.

---

*[gossip/client.go, line 301 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YCrryAaXgNbNiGPhe:-K9YLtbY_gsahsBLNSFZ:1464502251) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d984c0c7375b7fc82c7f3398b95b88a75cc2ad1/gossip/client.go#L301)):*
Yeah, we're arguing semantics. You're calling it "constructing `stream`" but it's really making an outbound connection. That's what I mean when I say that the context is passed to the outbound RPC.

---

*[gossip/server.go, line 134 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XsGoOcPpvAKGNfjFv:-K9YMD2gUOteU4ac-VYz:-1433356160) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L134)):*
Pushed a commit that does this. Will wait for @spencerkimball's rationale though.

---

*[gossip/server.go, line 204 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuZAHjqodwXaGK13F:-K9YLfwKfRnUF4Dgo5rc:-1843754257) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L204)):*
Changed this back. We don't want to set `args` to `nil` when we finally hit the error. There's a defer way up top that closes over `args` for cleanup.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 301 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YCrryAaXgNbNiGPhe:-K9YdRWPFWru1tBD5z8y:-1147319261) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d984c0c7375b7fc82c7f3398b95b88a75cc2ad1/gossip/client.go#L301)):*
I'm not trying to argue semantics but clarity. I was genuinely trying to figure out where that outbound RPC is - It's technically fair to call `stream` a (streaming) RPC, and thus to call constructing stream having an "outgoing RPC", but I would think of an outbound RPC as an invocation of `Send`, of which there is none here. My initially suggested `s/a/b/` reduces that source of confusion.

---

*[gossip/server.go, line 204 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuZAHjqodwXaGK13F:-K9Yer44ncUjtcT3p08S:-1121636583) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L204)):*
that defer looks like it should just grab the values right there and not close over `args`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Added a commit that moves the nodeID checking into the `for` loop; this should be better-behaved in the case of the remote peer bootstrapping (having node ID zero on the first request).

---

Review status: 13 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 301 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YCrryAaXgNbNiGPhe:-K9YjDQt6CTRvRrJ09ni:1336915767) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d984c0c7375b7fc82c7f3398b95b88a75cc2ad1/gossip/client.go#L301)):*
OK, done.

---

*[gossip/server.go, line 204 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuZAHjqodwXaGK13F:-K9Yj3LHFO83JHohkI47:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L204)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-:-K9Yjkpy-uI75fKUXZNm:-72889109)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[gossip/server.go, line 204 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuZAHjqodwXaGK13F:-K9YkgruM4oOcdeygSzd:-1973253918) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L204)):*
nit: now the `recvArgs` thing should be able to go in again.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[gossip/server.go, line 204 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuZAHjqodwXaGK13F:-K9YlZw57QCTPdcOzPD9:1988365492) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L204)):*
The RunWorker above still closes over `args` and wants it to be non-nil.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[gossip/server.go, line 204 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuZAHjqodwXaGK13F:-K9YnBiwA1g3dDj1zdbz:1228849309) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L204)):*
oh, now that I see this, can you explain why this isn't a race? You have the worker goroutine grabbing `args`, but you're also writing to it below and I can't see synchronization between the two.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[gossip/server.go, line 204 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuZAHjqodwXaGK13F:-K9YnjjWdjMzNhirzeRQ:-1340239541) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L204)):*
Everything is synchronized under `s.mu`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[gossip/server.go, line 204 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XuZAHjqodwXaGK13F:-K9YoeKOMwscQ0kC565k:1746563105) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L204)):*
ah, duh. Carry on.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird committed:
gossip: avoid local variable

bdarnell commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 139 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YewhpSET1i3bO6_cw:-K9YewhpSET1i3bO6_cx:1061765466) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/gossip/client.go#L139)):*
I don't understand. Why do we need to hold `g.mu` through the call to `stream.Send` but release it in the middle? Why can't we release the lock after constructing the request?

---

*[util/net.go, line 29 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Yct0VLBDnES4iKMYc:-K9Yct0WS--2-R46DL-t:67551489) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/util/net.go#L29)):*
I'd rather put these in a subpackage like `util/grpcutil` so we don't add the dependencies to our base `util` package. May not make much difference in the end, but it feels cleaner. (and then if it's in a subpackage maybe you can move `util/log/grpc.go` there too)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 139 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YewhpSET1i3bO6_cw:-K9Z5ZvDqFQ8g8g7HkND:522378880) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/gossip/client.go#L139)):*
Yeah, easy enough to remove these in the client. Done. It's providing more convenience in the server during calls to `Recv`, so I'm going to leave that one.

---

*[util/net.go, line 29 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Yct0VLBDnES4iKMYc:-K9Z5oTCmJDWxIxMQG-S:389159447) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/util/net.go#L29)):*
The parts of this that use grpc also depend on the unexported `listen` function, and exporting that function doesn't seem better.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 12 of 15 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[gossip/server.go, line 134 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9XsGoOcPpvAKGNfjFv:-K9bW3cWWUY_mysR1OFn:-1440580226) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d1c623ccecdfc9af2c739e3b4c52338e1a5418/gossip/server.go#L134)):*
I'm OK with it if you got it to work.

---

*[gossip/server.go, line 146 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9bWfWtlTd0ifTYh4Lg:-K9bWfWtlTd0ifTYh4Lh:-1322546420) ([raw file](https://github.com/cockroachdb/cockroach/blob/3786774be41a616c6b27d70065bcf6a3881526bd/gossip/server.go#L146)):*
It's problematic to do this in the inner loop. This rearrangement is now returning information to the client (via the RunTask above) even when the connection will be rejected by this code here when we determine there's no space for it and return an alternate address. This will completely screw up the gossip client's ability to know how far away (in terms of hops) non-peer nodes really are.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 12 of 15 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[gossip/server.go, line 146 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9bWfWtlTd0ifTYh4Lg:-K9bXxH4_vTG2cJa_kNd:1889443496) ([raw file](https://github.com/cockroachdb/cockroach/blob/3786774be41a616c6b27d70065bcf6a3881526bd/gossip/server.go#L146)):*
It's important to remember that once this connection is rejected, the Run{Worker,Task} above will terminate, because this function will have returned, which will tear down the stream. There's even no race, because the mutex is held throughout.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 3 files at r8, 4 of 4 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[gossip/server.go, line 146 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9bWfWtlTd0ifTYh4Lg:-K9bvO1eFGVSU-ye8ulw:1133336933) ([raw file](https://github.com/cockroachdb/cockroach/blob/3786774be41a616c6b27d70065bcf6a3881526bd/gossip/server.go#L146)):*
it'll print an error from `Send()` above in that case though, won't it? That seems dirty. Can the worker be started a tad later, after this logic here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 139 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YewhpSET1i3bO6_cw:-K9bqhE-kIQIkBYv62gk:-891949786) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/gossip/client.go#L139)):*
s/reply/request/

---

*[gossip/server.go, line 86 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9cFOa3EfAeXby9DNI6:-K9cFOa3EfAeXby9DNI7:332719980) ([raw file](https://github.com/cockroachdb/cockroach/blob/cca3900b01fdc21c05deda3b715cfb82907e4798/gossip/server.go#L86)):*
This is still too magical for my taste (I recognize that it's analogous to `sync.Cond.Wait` but that's not an API design that I would repeat). I'd rather make the unlock/lock pairs around each send/recv call explicit (and ideally make the critical sections smaller than that) 

---

*[util/net.go, line 29 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Yct0VLBDnES4iKMYc:-K9cJChdp7MxV9I0lHee:-1130001535) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/util/net.go#L29)):*
Why not? `listen()` looks like a reasonable thing to expose. Or we could move this whole file to package `util/net`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 139 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YewhpSET1i3bO6_cw:-K9cJzREKmRO207J34IG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/gossip/client.go#L139)):*
Done.

---

*[gossip/server.go, line 146 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9bWfWtlTd0ifTYh4Lg:-K9cPoK01LNYVwis9QSr:-170516206) ([raw file](https://github.com/cockroachdb/cockroach/blob/3786774be41a616c6b27d70065bcf6a3881526bd/gossip/server.go#L146)):*
That doesn't work, unfortunately. The worker above is the only thing that actually send the server's gossip to the client, so when a client comes in with NodeID `0` (bootstrapping), we need to send it the first range descriptor, or else it can't bootstrap, and everything hangs.

---

*[gossip/server.go, line 86 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9cFOa3EfAeXby9DNI6:-K9cOQNwIFpV7pD3pJyR:-1697134856) ([raw file](https://github.com/cockroachdb/cockroach/blob/cca3900b01fdc21c05deda3b715cfb82907e4798/gossip/server.go#L86)):*
Reducing the critical section is out of the scope of this PR.

This result of removing `unlockingServerStream` is going to be pretty ugly; we have a deferred unlock at the top of the function, which means we're going to have to do something like:
```
mu.Unlock()
err := stream.Send(...)
mu.Lock()
return err
```

This may be less magical, but it's certainly more error prone and tedious. Can we leave the code golf to future work? There's enough disagreement here that I don't think blocking this PR is warranted.

---

*[util/net.go, line 29 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Yct0VLBDnES4iKMYc:-K9cMg0VAlxTBQxJIL9G:1986329535) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/util/net.go#L29)):*
OK, I moved this to `util/grpcutil`.

Note that our `util` package will still depend on `http2`, and that `util/log/grpc.go` is unchanged, since `logDepth` is unexported.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 8 of 8 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 139 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YewhpSET1i3bO6_cw:-K9c_GZtGhKd9AK6BJFc:-1941035783) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/gossip/client.go#L139)):*
partial comment here.

---

*[gossip/server.go, line 146 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9bWfWtlTd0ifTYh4Lg:-K9cZbIEtUGa2nWENr9o:1773661908) ([raw file](https://github.com/cockroachdb/cockroach/blob/3786774be41a616c6b27d70065bcf6a3881526bd/gossip/server.go#L146)):*
what do we lose if the worker gets started after the `args.NodeID != 0` block? (ok, you need a `sync.Once`, but connections should be rare enough so that this cost isn't a factor). Or if you can think of another way to not have a confusing error logged in the event of a full node, I'm open for any reasonable way of getting this.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[gossip/server.go, line 86 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9cFOa3EfAeXby9DNI6:-K9cYqTT3Q8VDmjX5B0V:-67994563) ([raw file](https://github.com/cockroachdb/cockroach/blob/cca3900b01fdc21c05deda3b715cfb82907e4798/gossip/server.go#L86)):*
I feel pretty strongly about this; the best way to get concurrency right is to avoid magic and subtlety.  I find the quoted pattern preferable to `unlockingServerStream`. 

You've already changed the scope of the critical sections: each loop iteration used to be one critical section (with the lock released in `ready.Wait`); now the lock is released and reacquired three times per iteration (`Send`, `cycler.Wait`, `Recv`).  I think this needs to be visible instead of hidden in methods that do not appear to have anything to do with synchronization.

---

*[gossip/server.go, line 91 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9caUVkc0x7QQHivoH7:-K9caUVkc0x7QQHivoH8:-746883321) ([raw file](https://github.com/cockroachdb/cockroach/blob/19cfeca92db9d96ddfe5059055af04834c539cc5/gossip/server.go#L91)):*
Comment that `args` is accessed concurrently by this method and the `Worker`, and is guarded by `g.mu`. Consider moving the worker closure to a separate named function so we can be explicit about which local variables are shared 

---

*[gossip/server.go, line 204 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9cdyrlXgeL9P0ONM_8:-K9cdyrmLHdN-XIVGYC9:7303590) ([raw file](https://github.com/cockroachdb/cockroach/blob/19cfeca92db9d96ddfe5059055af04834c539cc5/gossip/server.go#L204)):*
This comment should also explain why `args` and `recvArgs` are two different variables.

---

*[util/net.go, line 29 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Yct0VLBDnES4iKMYc:-K9cfXLb6ftPe-uKu6Jn:-883176345) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/util/net.go#L29)):*
OK on the dep from `util` to `http2`. For the logging, do you need to call `logDepth`? In [storage/raft.go](https://github.com/cockroachdb/cockroach/blob/master/storage/raft.go)` I did a similar logging bridge using only public methods.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird committed:
supress uninteresting errors

tamird commented:
Review status: 14 of 16 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 139 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9YewhpSET1i3bO6_cw:-K9ceJKVH-EL0hD4zz6t:478069937) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/gossip/client.go#L139)):*
Derp, didn't mean to post anything here; I updated this as Ben suggested.

---

*[gossip/server.go, line 146 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9bWfWtlTd0ifTYh4Lg:-K9ceRkBQHy3y7p1K09p:-845929446) ([raw file](https://github.com/cockroachdb/cockroach/blob/3786774be41a616c6b27d70065bcf6a3881526bd/gossip/server.go#L146)):*
Added a thing that filters out "we hung up" errors.

---

*[gossip/server.go, line 86 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9cFOa3EfAeXby9DNI6:-K9coW8W6_xSxEbOODDR:-1876839553) ([raw file](https://github.com/cockroachdb/cockroach/blob/cca3900b01fdc21c05deda3b715cfb82907e4798/gossip/server.go#L86)):*
The scope of the critical section is not changed. The previous loop excluded `Send` and `Recv` by virtue of those methods being managed outside the scope of the function.

Your stamina exceeds mine. I've made the change.

---

*[gossip/server.go, line 91 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9caUVkc0x7QQHivoH7:-K9cn3c9LLpMusNNM7lx:-337249951) ([raw file](https://github.com/cockroachdb/cockroach/blob/19cfeca92db9d96ddfe5059055af04834c539cc5/gossip/server.go#L91)):*
Moving the worker to a separate function would prevent closing over `args` - or are you suggesting overwriting the pointed-to memory?

---

*[gossip/server.go, line 204 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9cdyrlXgeL9P0ONM_8:-K9cmw4KgsBWSXwSwbt-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/19cfeca92db9d96ddfe5059055af04834c539cc5/gossip/server.go#L204)):*
Done.

---

*[util/net.go, line 29 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9Yct0VLBDnES4iKMYc:-K9clY1hEcme1HDGIRNh:-2048594082) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ca9e2d5853d6b6323dc5f21e48c1bd3bbda0a5/util/net.go#L29)):*
You're right. Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 13 of 16 files reviewed at latest revision, 6 unresolved discussions.

---

*[gossip/server.go, line 91 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9caUVkc0x7QQHivoH7:-K9cqhpt12dYbVebyk0T:971796679) ([raw file](https://github.com/cockroachdb/cockroach/blob/19cfeca92db9d96ddfe5059055af04834c539cc5/gossip/server.go#L91)):*
I'm suggesting passing `&args` to the worker function instead of letting it get picked up implicitly by the closure (to make sure that `args` is the only variable that gets shared and we don't end up with both threads trying to overwrite the same `err` variable or something) 

```
func (s *server) gossipWorker(args *ReturnTypeOfRecv) {
}

...
    s.stopper.RunWorker(func() {
        s.gossipWorker(&args)
    })
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 13 of 16 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[gossip/server.go, line 91 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-K9caUVkc0x7QQHivoH7:-K9cvMV3ttmfSRNEXcZk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/19cfeca92db9d96ddfe5059055af04834c539cc5/gossip/server.go#L91)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 4 files at r11.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Status update: the upstream PR is being broken up so it's easier to review and merge. Hopefully it lands tomorrow or the next day.

tamird commented:
Upstream PR landed, I'll be merging this shortly.

tschottdorf commented:
:rage:

On Thu, Feb 11, 2016 at 10:45 PM Tamir Duberstein <notifications@github.com>
wrote:

> Upstream PR landed, I'll be merging this shortly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4080#issuecomment-183169394>
> .
>

tamird commented:
Well, I was wrong. Data races in upstream, see https://github.com/grpc/grpc-go/issues/551

tamird committed:
gossip: synchronize Send with function return

Yuck.

tamird commented:
Added a commit that fixed the data races; turns out it is not safe to `Send()` concurrently with other `Send` calls _and_ with returning from the function.

bdarnell commented:
Review status: 5 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

*[gossip/server.go, line 116 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-KAMrRP725r_3VHyFyXA:-KAMrRP725r_3VHyFyXB:284584418) ([raw file](https://github.com/cockroachdb/cockroach/blob/48db7611aeb0405a0ab97c531fad61cafe489b46/gossip/server.go#L116)):*
Won't this deadlock if this branch is taken, since nothing will release the syncChan?

Even if the `defer release()` were moved up so it were called on `<-ctx.Done()`, I'm having a hard time convincing myself that this is correct. I think I'd split this up to use both a `Mutex` (to keep `Send` calls from running concurrently with each other) and a `WaitGroup` (to guard the exit from the function).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 5 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

*[gossip/server.go, line 116 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-KAMrRP725r_3VHyFyXA:-KAMshw_ork4eNPcmODe:-46382459) ([raw file](https://github.com/cockroachdb/cockroach/blob/48db7611aeb0405a0ab97c531fad61cafe489b46/gossip/server.go#L116)):*
The syncChan will not have been acquired in that case, so no release is necessary.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 5 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

*[gossip/server.go, line 116 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-KAMrRP725r_3VHyFyXA:-KAMtBcm1uGC5f4IpUjV:-476461518) ([raw file](https://github.com/cockroachdb/cockroach/blob/48db7611aeb0405a0ab97c531fad61cafe489b46/gossip/server.go#L116)):*
Ah, I see.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4080#-:-KAMtDdicT4fWLf-0Cg6:676834233)*
<!-- Sent from Reviewable.io -->
